File: leetcode/src/python/balanced_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class BalancedBinaryTree:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        else:
            is_left_balanced = self.isBalanced(root.left)
            is_right_balanced = self.isBalanced(root.right)
            if is_left_balanced and is_right_balanced:
                left_depth = self.maxDepth(root.left)
                right_depth = self.maxDepth(root.right)
                return abs(left_depth - right_depth) <= 1
            else:
                return False
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        result = 0
        if root != None:
            left_depth = 1 + self.maxDepth(root.left)
            right_depth = 1 + self.maxDepth(root.right)
            result = max(left_depth, right_depth)
        return result

File: leetcode/src/scala/BalancedBinaryTree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object BalancedBinaryTree2 {
    def isBalanced(root: TreeNode): Boolean = {
        if (root == null) return true
        else return isBalanced(root.left) && isBalanced(root.right) && Math.abs(maxDepth(root.left) - maxDepth(root.right)) <= 1
    }
    def maxDepth(root: TreeNode): Int = {
        if (root == null) return 0
        else return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1
    }
    def main(args : Array[String]): Unit = {
        println(isBalanced(new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)))))
    }
}